### PR TITLE
`ultralytics 8.0.239` Ultralytics Actions and `hub-sdk` adoption

### PR DESCRIPTION
**Pull request by glenn-jocher**
This pull request adds an improved Ultralytics Actions workflow to automatically format code and documentation to the new Ultralytics official standards maintained at https://github.com/ultralytics/actions.

**Five** individual actions are run by default now including a **new broken links check for markdown and HTML files**. Disable individual actions by setting them to false or removing their line, i.e. delete 'markdown: true' line to disable markdown formatting.

Note that additional spellings have been added to the Ultralytics spelling dictionary and frontmatter is now ignored by markdown formatters per your feedback.

To customize an action use a `pyproject.toml` file in this repo. For details see https://github.com/ultralytics/actions. 

```yaml
# Ultralytics 🚀 - AGPL-3.0 license
# Ultralytics Actions https://github.com/ultralytics/actions
# This workflow automatically formats code and documentation in PRs to official Ultralytics standards

name: Ultralytics Actions

on:
  push:
    branches: [main]
  pull_request:
    branches: [main]

jobs:
  format:
    runs-on: ubuntu-latest
    steps:
      - name: Run Ultralytics Formatting
        uses: ultralytics/actions@main
        with:
          token: ${{ secrets.GITHUB_TOKEN }}  # automatically generated
          python: true
          docstrings: true
          markdown: true
          spelling: true
          links: true
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

**WARNING ⚠️** this PR is very large, summary may not cover all changes.

### 🌟 Summary
Enhanced Python version support for YOLOv8 continuous integration workflows.

### 📊 Key Changes
- Python version for CI workflows updated to `3.11`
- New format workflow file added for automated code and documentation formatting
- Unnecessary double-quote string fixer removed from pre-commit configuration
- Updated documentation build scripts for clarity and consistency in string quoting
- Adjusted Python scripts under 'examples' directory with string quoting and path handling improvements
- Documentation update script enhanced with better handling of HTML links in markdown
- Revised build reference and translation update scripts for cleaner code organization
- Minor code adjustments in data splitting script for DOTA dataset 

### 🎯 Purpose & Impact 
- Ensures compatibility with the latest Python release, maintaining robust testing across newer environments
- Automates code style enforcement, leading to a more unified and readable codebase
- Streamlines documentation generation, resulting in more reliable updates to user-facing content
- Improves example scripts, making them easier to read and follow, increasing their educational value
- Enhances internal processing scripts, making maintenance and updates simpler, and reducing potential errors
- The overall impact is an increased quality and reliability of the Ultralytics software offerings, benefiting both developers and end-users.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@Burhan-Q I see it keeps failing on a random link check failure, I'm going to disable link checks here (they're still in their own action on a schedule in links.yml)

**Comment by glenn-jocher** [comment]
@Burhan-Q I've merged @AyushExel's docs PR into main, this is the last PR left other than @kalenmike's HUB-SDK PR.

**Comment by Burhan-Q** [comment]
I'm about 1/2 way thru the review 
image

All the comments that I found that looked like they should get moved have been moved tho.

**Comment by glenn-jocher** [comment]
@Burhan-Q ok awesome. I'm going to go checkout @kalenmike PR and try to get that in now and see what's up with the HUB CI fail on main branch now. 

Do you know if the server is expecting the new PR to be merged for new trainings as of now?

**Comment by Burhan-Q** [comment]
@glenn-jocher I'm not sure. I was only ever testing on the staged environment. The link at on the last line makes me think it's getting a `model_id` value of `***` since it's usually a string of characters and numbers

**Comment by glenn-jocher** [comment]
Oh haha the asterixes are just GitHub actions obscuring keys and passwords etc, it's getting the actual key, but the console won't show it to you for security

**Comment by Burhan-Q** [comment]
That makes sense, better to obscure it even if it's not sensitive that to leak when something is sensitive

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -299,19 +327,23 @@ def _do_train(self, world_size=1):
...
-                    f"Logging results to {colorstr('bold', self.save_dir)}\n"
-                    f'Starting training for '
-                    f'{self.args.time} hours...' if self.args.time else f'{self.epochs} epochs...')
+        self.run_callbacks("on_train_start")
+        LOGGER.info(
+            f'Image sizes {self.args.imgsz} train, {self.args.imgsz} val\n'
+            f'Using {self.train_loader.num_workers * (world_size or 1)} dataloader workers\n'
+            f"Logging results to {colorstr('bold', self.save_dir)}\n"
+            f'Starting training for '
+            f'{self.args.time} hours...'
```
Comment (Burhan-Q) [inline ultralytics/engine/trainer.py:336]: These can get changed to double quotes. Wonder if the new line/indentation layout caused the quotes replacement get bypassed.

File: ultralytics/engine/trainer.py [hunk 2]
```diff
@@ -661,41 +705,45 @@ def build_optimizer(self, model, name='auto', lr=0.001, momentum=0.9, decay=1e-5
...
+            )
 
-        optimizer.add_param_group({'params': g[0], 'weight_decay': decay})  # add g0 with weight_decay
-        optimizer.add_param_group({'params': g[1], 'weight_decay': 0.0})  # add g1 (BatchNorm2d weights)
+        optimizer.add_param_group({"params": g[0], "weight_decay": decay})  # add g0 with weight_decay
+        optimizer.add_param_group({"params": g[1], "weight_decay": 0.0})  # add g1 (BatchNorm2d weights)
         LOGGER.info(
             f"{colorstr('optimizer:')} {type(optimizer).__name__}(lr={lr}, momentum={momentum}) with parameter groups "
-            f'{len(g[1])} weight(decay=0.0), {len(g[0])} weight(decay={decay}), {len(g[2])} bias(decay=0.0)')
+            f'{len(g[1])} weight(decay=0.0), {len(g[0])} weight(decay={decay}), {len(g[2])} bias(decay=0.0)'
```
Comment (Burhan-Q) [inline ultralytics/engine/trainer.py:747]: Could get updated to use `"` quotes

File: ultralytics/models/yolo/pose/val.py [hunk 3]
```diff
@@ -187,37 +208,41 @@ def pred_to_json(self, predn, filename):
...
                         eval.params.imgIds = [int(Path(x).stem) for x in self.dataloader.dataset.im_files]  # im to eval
                     eval.evaluate()
                     eval.accumulate()
                     eval.summarize()
                     idx = i * 4 + 2
-                    stats[self.metrics.keys[idx + 1]], stats[
-                        self.metrics.keys[idx]] = eval.stats[:2]  # update mAP50-95 and mAP50
+                    stats[self.metrics.keys[idx + 1]], stats[self.metrics.keys[idx]] = eval.stats[
+                        :2
+                    ]  # update mAP50-95 and mAP50
```
Comment (Burhan-Q) [inline ultralytics/models/yolo/pose/val.py:245]: This seems _excessive_ to me

File: ultralytics/nn/modules/block.py [hunk 4]
```diff
@@ -284,9 +302,11 @@ def __init__(self, c1, c2, k=3, s=1):
         self.conv = nn.Sequential(
             GhostConv(c1, c_, 1, 1),  # pw
             DWConv(c_, c_, k, s, act=False) if s == 2 else nn.Identity(),  # dw
-            GhostConv(c_, c2, 1, 1, act=False))  # pw-linear
-        self.shortcut = nn.Sequential(DWConv(c1, c1, k, s, act=False), Conv(c1, c2, 1, 1,
-                                                                            act=False)) if s == 2 else nn.Identity()
+            GhostConv(c_, c2, 1, 1, act=False),
+        )  # pw-linear
```
Comment (Burhan-Q) [inline ultralytics/nn/modules/block.py:306]: Move comment to L305

File: ultralytics/nn/modules/transformer.py [hunk 5]
```diff
@@ -339,14 +353,16 @@ def forward(self, embed, refer_bbox, feats, shapes, padding_mask=None, attn_mask
 
         # Self attention
         q = k = self.with_pos_embed(embed, query_pos)
-        tgt = self.self_attn(q.transpose(0, 1), k.transpose(0, 1), embed.transpose(0, 1),
-                             attn_mask=attn_mask)[0].transpose(0, 1)
+        tgt = self.self_attn(q.transpose(0, 1), k.transpose(0, 1), embed.transpose(0, 1), attn_mask=attn_mask)[
+            0
```
Comment (Burhan-Q) [inline ultralytics/nn/modules/transformer.py:357]: This seems _excessive_ to me

**Comment by Burhan-Q** [comment]
@glenn-jocher 
image

**Review by Burhan-Q — APPROVED** [review]
All notes for future reference, nothing critical to fix

File: ultralytics/data/augment.py [hunk 6]
```diff
@@ -194,15 +195,15 @@ def _mosaic3(self, labels):
...
+            img3[y1:y2, x1:x2] = img[y1 - padh :, x1 - padw :]  # img3[ymin:ymax, xmin:xmax]
             # hp, wp = h, w  # height, width previous for next iteration
 
             # Labels assuming imgsz*2 mosaic size
             labels_patch = self._update_labels(labels_patch, padw + self.border[0], padh + self.border[1])
             mosaic_labels.append(labels_patch)
         final_labels = self._cat_labels(mosaic_labels)
 
-        final_labels['img'] = img3[-self.border[0]:self.border[0], -self.border[1]:self.border[1]]
+        final_labels["img"] = img3[-self.border[0] : self.border[0], -self.border[1] : self.border[1]]
```
[the remark this answers is not in the document]
Comment (glenn-jocher) [reply ultralytics/data/augment.py:206]: @Burhan-Q yeah the spacing just is what it is with the default Black/Ruff formatters. Not sure if I like it either but we can update in the future too.

File: ultralytics/data/base.py [hunk 7]
```diff
@@ -103,43 +105,43 @@ def get_img_files(self, img_path):
...
+            im_files = sorted(x.replace("/", os.sep) for x in f if x.split(".")[-1].lower() in IMG_FORMATS)
             # self.img_files = sorted([x for x in f if x.suffix[1:].lower() in IMG_FORMATS])  # pathlib
-            assert im_files, f'{self.prefix}No images found in {img_path}'
+            assert im_files, f"{self.prefix}No images found in {img_path}"
         except Exception as e:
-            raise FileNotFoundError(f'{self.prefix}Error loading data from {img_path}\n{HELP_URL}') from e
+            raise FileNotFoundError(f"{self.prefix}Error loading data from {img_path}\n{HELP_URL}") from e
         if self.fraction < 1:
-            im_files = im_files[:round(len(im_files) * self.fraction)]
+            im_files = im_files[: round(len(im_files) * self.fraction)]
```
[the remark this answers is not in the document]
Comment (glenn-jocher) [reply ultralytics/data/base.py:124]: I know, it's odd, but changeable if we want with a setting.

File: ultralytics/nn/modules/block.py (hunk 4 above)
Comment (glenn-jocher) [reply ultralytics/nn/modules/block.py]: Done!

File: ultralytics/data/converter.py [hunk 8]
```diff
@@ -20,10 +20,98 @@ def coco91_to_coco80_class():
...
+        None,
+        73,
+        74,
+        75,
+        76,
+        77,
+        78,
+        79,
+        None,
+    ]
```
[the remark this answers is not in the document]
Comment (glenn-jocher) [reply ultralytics/data/converter.py:114]: Yeah what we need to do is look at the Ruff options, I'm sure there's something to wrap lists like this.

File: ultralytics/models/fastsam/prompt.py [hunk 9]
```diff
@@ -258,18 +264,18 @@ def _crop_image(self, format_results):
...
         filter_id = []
         for _, mask in enumerate(annotations):
-            if np.sum(mask['segmentation']) <= 100:
+            if np.sum(mask["segmentation"]) <= 100:
                 filter_id.append(_)
                 continue
-            bbox = self._get_bbox_from_mask(mask['segmentation'])  # mask 的 bbox
+            bbox = self._get_bbox_from_mask(mask["segmentation"])  # mask 的 bbox
             cropped_boxes.append(self._segment_image(image, bbox))  # 保存裁剪的图片
             cropped_images.append(bbox)  # 保存裁剪的图片的bbox
```
[the remark this answers is not in the document]
Comment (glenn-jocher) [reply ultralytics/models/fastsam/prompt.py:280]: Oh haha yes

[the remark this answers is not in the document]
Comment (glenn-jocher) [reply ultralytics/models/fastsam/prompt.py]: Done!

File: ultralytics/utils/ops.py [hunk 10]
```diff
@@ -76,9 +76,15 @@ def segment2box(segment, width=640, height=640):
     # Convert 1 segment label to 1 box label, applying inside-image constraint, i.e. (xy1, xy2, ...) to (xyxy)
     x, y = segment.T  # segment xy
     inside = (x >= 0) & (y >= 0) & (x <= width) & (y <= height)
-    x, y, = x[inside], y[inside]
-    return np.array([x.min(), y.min(), x.max(), y.max()], dtype=segment.dtype) if any(x) else np.zeros(
-        4, dtype=segment.dtype)  # xyxy
+    (
+        x,
+        y,
+    ) = x[inside], y[inside]
```
[the remark this answers is not in the document]
Comment (glenn-jocher) [reply ultralytics/utils/ops.py]: +1 fixed

**Pull request merged**